bsip-0044.md

```diff
@@ -156,9 +167,101 @@ Alice now examines the HTLC Bob created on the Bitcoin Network, ensuring the `pr
+
+  // Constructor
+  Set: id = next available
+  Initialize other fields = NULL
+
+  return contract
+```
+
+## **Operations**
+
+### **Prepare**
+
+```
+transaction_obj htlc_prepare(depositor, quantity, symbol, recipient, preimage_hash, preimage_length, timeout_threshold, htlc_preparation_fee)
+  Validate: HTLC signed by requisite `authority` for `depositor` account 
+  Validate: `depositor` account has requisite `quantity` of `symbol` asset for the `guarantee`
+  Validate: `timeout_threshold` < now() + GRAPHENE_HTLC_MAXIMUM_DURRATION
+  Calculate: `required_fee` = GRAPHENE_HTLC_OPERATION_FEE + GRAPHENE_HTLC_DAILY_FEE * count((`timeout_threshold` - now()), days)
+  Validate: `depositor` account has requisite `quantity` of BTS for `required_fee`
+  Validate: `recipient` account exists
+  Validate: `preimage_length` does not exceed GRAPHENE_HTLC_MAXIMUM_PREIMAGE_LENGTH
+  Validate: `preimage_hash` well formed
+  Update: BTS balance of `depositor` based on `required_fee`)
+  contract = new htlc_obj
+  Set: `contract.depositor` = `depositor`
+  Set: `contract.recipient` = `recipient`
+  Set: `contract.preimage_hash` = `preimage_hash`
+  Set: `contract.preimage_length` = `preimage_length`
+  Set: `contract.timeout_treshold` = `timeout_threshold`
+  Transfer: from `depositor` account to `contract.quantity` of `contract.symbol`
+  return results
+```
+
+### **Redeem**
+
+```
+transaction_obj htlc_redeem(redeemer, id, preimage, htlc_redemption_fee)
+  Validate: transaction signed by requisite `authority` for `redeemer` // any account may attempt to redeem
+  Get: get_htlc(id)
+  Validate: `redeemer` account has requisite `quantity` of BTS for `htlc_redeem_fee` and `htlc_kb_fee`
+  // Evaluate: timelock
+     if now() < `timeout_threshold` then return error // "timeout exceeded"
+  Update: balance of `redeemer` based on total fees
+  // Evaluate: hashlock
+     if length(preimage) != `id.preimage_length` then return error // "preimage length mismatch"
+     Calculate: `preimage_hash` = hash(preimage)
+     if `preimage_hash` != `id.preimage_hash` then return error // "invalid preimage submitted"
+     Update: balance of `id.recipient` add asset `id.symbol` of quantity `id.quantity`
+     Add: transaction to mempool 
+     Set: `id.preimage_tx_id` = `transaction_id` 
+     Cleanup: memory allocated to this htlc
+     return: results
+```
+
+### **Extend Expiry**
+
+```
+transaction_obj htlc_extend_expiry(depositor, id, timeout_threshold, htlc_extention_fee)
+  Validate: transaction signed by requisite `authority` for `depositor` account 
+  Validate: `timeout_threshold` < now() + GRAPHENE_HTLC_MAXIMUM_DURRATION
+  Calculate: `required_fee` = GRAPHENE_HTLC_DAILY_FEE * count((`timeout_threshold` - now()), days)
+  Validate: `depositor` account has requisite `quantity` of BTS for `required_fee`
+  Update: BTS balance of `depositor` based on `required_fee`)
+  Set: `contract.timeout_treshold` = `timeout_threshold`
+  return results
+```
+
```

Comment:
Please add a virtual operation for redemption containing (depositor, recipient, amount, symbol, preimage) and one for expiration containing (depositor, amount, symbol). (Perhaps include HTLC object id in both? Unsure.)

Reply:
@jmjatlanta you gotta help with virtual ops, as that is beyond my ability to spec out properly.

Reply:
If I understand virtual operations correctly, we need a non-virtual operation for redemption, as the preimage must be stored on the chain. Expiration does not affect the chain (only account balances), so should be a virtual operation. Please correct me if I am wrong.

Reply:
You have already correctly specified a non-virtual operation for redemption. Per default, that operation would only turn up in the account history of the redeemer, not in that of the depositor. IMO the depositor should see the outcome of the htlc in his history, so we either need an extra virtual op for this, or have to make sure that the non-virtual redemption op turns up in the depositor's history (which is probably simpler).
Expiration should appear in the depositor's history and is not otherwise represented, therefore a virtual op is required.

Reply:
I've attempted to reflect the virtual operations @pmconrad mentioned with the addition of `html_at_expiry` and inline within `htlc_redeem`. However, I'll need some guidance on both logic and format.